src/Npgsql/NpgsqlDatabaseInfo.cs

```diff
@@ -15,8 +15,8 @@ public abstract class NpgsqlDatabaseInfo
     {
         #region Fields
 
-        internal static readonly ConcurrentDictionary<string, NpgsqlDatabaseInfo> Cache
-            = new ConcurrentDictionary<string, NpgsqlDatabaseInfo>();
+        internal static readonly ConcurrentDictionary<(string? Host, int Port, string? Database), NpgsqlDatabaseInfo> Cache
```

Comment:
Unfortunately, `NpgsqlConnectionStringBuilder` says that these fields are nullable. Do you want some bangs here? (:

Reply:
I hate bangs as much as you do :)

The thing to do here, is refactor things a bit. We already [validate](https://github.com/npgsql/npgsql/blob/8a479081f707784b5040747b23102c3d6371b9d3/src/Npgsql/NpgsqlConnectionStringBuilder.cs#L1496) that Host is never null, and we even have [a bang'ed non-nullable property](https://github.com/npgsql/npgsql/blob/adc96c81c7f52e945516327ab52da469ecfda7d0/src/Npgsql/NpgsqlConnector.cs#L337) because of this which you can use. It would be cleaner to remove the bang from that property by make it a non-nullable auto-property, and just assign the host after we've validated it.

For Database, we know it can't be null because [we null-coalesce Database to the user](https://github.com/npgsql/npgsql/blob/adc96c81c7f52e945516327ab52da469ecfda7d0/src/Npgsql/NpgsqlConnector.cs#L419). We could use the same kind of technique - have a bang'ed property, or more cleanly by having a non-nullable auto-property. Another option is to pass the coalesced, non-nullable database through the stack (this avoids adding another property), but it makes the code a bit nastier.

IMHO this is a good example of how the nullable references features forces us to clean up code to be better :)

Reply:
Will fix it, but there is a bigger problem as said above. It's more important problem than bangs.

Reply:
It's because of test run parallelization.